dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h

```diff
@@ -170,27 +170,32 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
     CUDA_CALL(cudaEventRecord(hw_decode_event_, hw_decode_stream_));
 
 #if NVJPEG2K_ENABLED
-    auto nvjpeg2k_thread_id = nvjpeg2k_thread_.GetThreadIds()[0];
-    if (device_memory_padding_jpeg2k > 0) {
-      // Adding smaller buffers that are allocated by nvjpeg2k on startup.
-      // The sizes were obtained empirically.
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU, 1024);
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU, 4 * 1024);
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU, 16 * 1024);
-      nvjpeg2k_intermediate_buffer_.resize(device_memory_padding_jpeg2k / 8);
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU,
-                               device_memory_padding_jpeg2k);
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU,
-                               device_memory_padding_jpeg2k);
-    }
-    if (host_memory_padding_jpeg2k > 0) {
-      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::Pinned,
-                               host_memory_padding_jpeg2k);
-    }
-    nvjpeg2k_thread_.AddWork([this](int) {
+
+    nvjpeg2k_thread_.AddWork([this, device_memory_padding_jpeg2k, host_memory_padding_jpeg2k](int) {
       nvjpeg2k_handle_ = NvJPEG2KHandle(&nvjpeg2k_dev_alloc_, &nvjpeg2k_pin_alloc_);
 
-      nvjpeg2k_decoder_ = NvJPEG2KDecodeState(nvjpeg2k_handle_);
+      // nvJPEG2k doesn't support deprecated sm while DALI does. In this case, NvJPEG2KHandle may
+      // be empty and we need to fall back to the CPU implementation
+      if (nvjpeg2k_handle_) {
+        auto nvjpeg2k_thread_id = nvjpeg2k_thread_.GetThreadIds()[0];
```

Comment:
I feel it'd be better to get the thread id before posting the work

Reply:
Done